Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp

```diff
@@ -120,52 +129,181 @@ bool CreateGlobalStateManager(Worker_SnapshotOutputStream* OutputStream)
+	FUnresolvedObjectsMap UnresolvedObjectsMap;
+	FUnresolvedObjectsMap HandoverUnresolvedObjectsMap;
+	ComponentFactory DataFactory(UnresolvedObjectsMap, HandoverUnresolvedObjectsMap, NetDriver);
+
+	// Create initial state of Actor (which is the state the Actor is in before running the level)
+	TArray<Worker_ComponentData> ComponentData = DataFactory.CreateComponentDatas(Actor, InitialRepChanges, InitialHandoverChanges);
+
+	// Actor RPCs
+	for (int RPCType = 0; RPCType < RPC_Count; RPCType++)
+	{
+		ComponentData.Add(ComponentFactory::CreateEmptyComponentData(Info->RPCComponents[RPCType]));
+	}
+
+	// Process Subobjects
+	for (UClass* SubobjectClass : Info->SubobjectClasses)
+	{
+		FClassInfo* ComponentInfo = TypebindingManager->FindClassInfoByClass(SubobjectClass);
+		check(ComponentInfo);
+
+		TArray<UObject*> DefaultSubobjects;
+		Actor->GetDefaultSubobjects(DefaultSubobjects);
+		UObject** FoundSubobject = DefaultSubobjects.FindByPredicate([SubobjectClass](const UObject* Obj)
+		{
+			return (Obj->GetClass() == SubobjectClass);
+		});
+		check(FoundSubobject);
+		UObject* Subobject = *FoundSubobject;
+
+		FRepChangeState SubobjectRepChanges = Channel->CreateInitialRepChangeState(Subobject);
+		FHandoverChangeState SubobjectHandoverChanges = Channel->CreateInitialHandoverChangeState(ComponentInfo);
+
+		// Initial state of subobject
+		ComponentData.Append(DataFactory.CreateComponentDatas(Subobject, SubobjectRepChanges, SubobjectHandoverChanges));
+
+		// Subobject RPCs
+		for (int RPCType = 0; RPCType < RPC_Count; RPCType++)
+		{
+			ComponentData.Add(ComponentFactory::CreateEmptyComponentData(ComponentInfo->RPCComponents[RPCType]));
+		}
+	}
+
+	return ComponentData;
+}
+
+bool CreateStartupActor(Worker_SnapshotOutputStream* OutputStream, AActor* Actor, Worker_EntityId EntityId, USpatialNetConnection* NetConnection, USpatialTypebindingManager* TypebindingManager)
+{
+	Worker_Entity Entity;
+	Entity.entity_id = EntityId;
+
+	UClass* ActorClass = Actor->GetClass();
+
+	FClassInfo* ActorInfo = TypebindingManager->FindClassInfoByClass(ActorClass);
+	check(ActorInfo);
+
+	WriteAclMap ComponentWriteAcl;
+	ComponentWriteAcl.Add(POSITION_COMPONENT_ID, UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ROTATION_COMPONENT_ID, UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->SingleClientComponent, UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->MultiClientComponent, UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->HandoverComponent, UnrealWorkerPermission);
+	// No write attribute for RPC_Client since a Startup Actor will have no owner on level start
+	ComponentWriteAcl.Add(ActorInfo->RPCComponents[RPC_Server], UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->RPCComponents[RPC_CrossServer], UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->RPCComponents[RPC_NetMulticast], UnrealWorkerPermission);
+
+	for (UClass* SubobjectClass : ActorInfo->SubobjectClasses)
```

Comment:
This code seems messy. Two different ways of iterating through actor subobjects.
Also worth consolidating the common parts of entity/component creation to some helper functions that can be accessed at runtime.

Reply:
Okay that's fine, but can we create a Jira ticket to track this.